### DISCUSSION TITLE
Reducing CompletableFuture allocations query-time

### DISCUSSION THREAD

**Discussion by rrva** [graphql-java/graphql-java#3175] [General]
When profiling graphql-java and focusing on query-time allocations, CompletableFuture comes out at the very top. This is due to that each resolved field allocates at least one CompletableFuture, regardless of if the DataFetcher is async or not, this is just the very nature of how AsyncExecutionStrategy is implemented. Also there seems to be many intermediate CompletableFuture objects allocated, where there are several thenApply() etc steps.

During a 5 minute load test using queries which are completely resolved locally in-process, using java flight recording, and allocation profiling enabled I get the following allocation stats (18,9% of all allocations are CompletableFuture instances).

<img width="1175" alt="image" src="https://user-images.githubusercontent.com/887132/230912612-07634d40-f468-4d48-83a3-17825b6eb58b.png">

Some ideas:

1. Rewrite AsyncExecutionStrategy and base classes to produce less CompletableFuture objects per resolved field. The overall API I guess needs to be CF based even for fields that can be resolved synchronously but at least the amount of temporary intermediate CFs could be reduced.

2. Since we run Kotlin, and I guess several other graphql-java users do, one possible area of improvement is to implement an ExecutionStrategy based on kotlin coroutines, which would need less CF objects being allocated at least internally inside the execution strategy.

What do you think?

**Comment by bbakerman** [comment]
> What do you think?

> Rewrite AsyncExecutionStrategy and base classes to produce less CompletableFuture objects per resolved field.

The challenge is that to be truly async - you need to embrace an async mechanism.  In our case with Java 8  as a minimum when we embraced an async design, we had CompletableFutures to embrace.

We do more than just fetch "field values" that are CFs - we compose functionality together.

For example

```
someField {
   fieldA
   fieldB
   fieldC
}
```

We need to asynchronously compose together those fields.  And CF.allOF() et al is how we do that.  If we had a list of values, where some are sync objects and some are not then we would make the code a nightmare in terms of trying to compose together steps we make.  

Other code like `Instrumentation` have an API that is CF based.  

In short CF is FUNDAMENTAL to how graphql-java works right now.  We really cant avoid it now given the choices we have made.

We would be breaking both API and semantics if we moved to a half sync values / half CFs say.

> queries which are completely resolved locally in-process

The fact is in a real world production system - resolving values in sync (I think you meant synchronously and in process) is not likely.    Useful real world apps hit databases and other services and these are inherently async mechanisms.

So yes in your test it feels pointless - but I posit it's not a real world application and hence exaggerates the drawbacks from the benefits.

> Since we run Kotlin, and I guess several other graphql-java users do, one possible area of improvement is to implement an ExecutionStrategy based on kotlin coroutines

This is doable.  However you would be on your own for now.   We have found that truly composeable ExecutionStratgeies are impossible - namely because the design choices are embedded into code.  A coroutines ES would be designed differently and should be.

This also has flow on effects to callbacks like `Instrumentation`.  The `Instrumentation` of today IS dictated by the behavior of the engine.  So if you right a new ES engine, you will also need new Instrumentation that is sympathetic to how you execute.


In the medium term future, there is JVM virtual threads.  Basically I think it will kill the async patterns INCLUDING (probably) coroutines as they exist today.

We have done some POCs on this and it involves a new ES engine as well.  So at some time graphql-java is likely to embrace virtual threads.  But we wont until they become fully GA and widely used.  And Java 19 is preview say.

**Reply by bbakerman** [reply]
As an aside - we do more with CompletableFutures and synchronous values than just wrap them

They allow us to return a value OR and exception but on the one representation

```
CompletableFuture cf = fetchSomethingSynchrnous(...)
```

The cf above can be a completed CF or it can be excetional.  Later layers of code can either use the value OR the exception but its represented in one returned value without lots of try catches


```
cf.handle( (value, exception) -> { .....})
```

**Reply by bbakerman** [reply]
I had a look into CFs a little more yesterday and their compoising mechanism will indeed create lots of object

```
CompletableFuture<X> cf = initialFetch();

return cf.thenApply( x -> someWorkonThat(x)).thenCompose( y -> someMoreWork(y)).thenApply( z -> twizzle(z));
```

Each `.then` step creates a new CF and this is pushed onto a stack of "completion stages" in a dependant order.  

So if you use `CompletableFuture` as you async composition mechanism - you will inherently end up with lots of CF objects being allocated.

The JVM is really good at managing short lived objects.  If an object never makes it out of the young GC pools (eden space) then its really cheap to create  and let go.  

in CF terms this then means how long between when a `.then` step takes to run.  If it's fast then the allocated CF wont make it out of eden space - if it takes time, then the CF object will mature and start to become more expensive in terms of GC and hence in terms of GC overhead.

So when we look at object allocation and performance - we need to consider not just the counts of objects - but their ternue.

**Reply by rrva** [reply]
Thanks for looking at this. I still believe that even the allocation of objects that die young can be a limiting factor on performance but I have not extensively measured this. I do see that GC activity is a large deal of total CPU time even after tuning GC (G1 GC on Java 20 with plenty of heap). Also, we have seen a marked performance increase when benchmarking a JVM with optimized GC on this workload (Azul Prime). My plan was to implement a execution strategy which needs less allocations and see if it performs better. In my specific case, I can forfeit the need to have instrumentations even and then the instrumentation contract with CFs is less important short term. Right now we only have an instrumentation that instruments the execution result for setting shorter HTTP Cache headers for certain queries. Possibly this can be done without instrumentations, and something that the data fetcher can supply via extensions. So my hypothesis is that allocations in query time is a limiting factor.

**Reply by bbakerman** [reply]
I do wish you good luck with it.  I have no doubt that a lighter weight and inherently faster engine could be written

By removing design constraints like Instrumentation (eg having to have to call it) or by streamlining what happens during field value fetch you will improve the speed for sure.

I am very interested in by how much.   eg current benchmarked against the future one.  This would be great dtaa to have.

**Comment by dfa1** [comment]
hello @rrva 

I started to tackle this problem too! 

Please have a look here https://github.com/graphql-java/graphql-java/pull/3233, it is not much but maybe a good starting point. I see also other point that could be improved:
- @bbakerman maybe this is a crazy idea but how about not using `CompletableFuture` in case of a `TrivialDataFetcher`? My hope is that since `PropertyDataFetcher ` is just doing in memory lookups, it does not really require async;
- allocations of `ExecutionStrategyParameters` + Builder (it is always 2 objects);
- `ExecutionStrategyParameters` has 2 fields that are not used anymore (`listSize` and `currentListIndex`): here it is hard to remove them because they are part of a public API class;
- @bbakerman  `NonNullableFieldValidator` is also allocated per field: maybe there is a way to just use parameter passing for it? unfortunately, it is part of the public API... so it will be hard to change (there is also another problem here: it is part of `ExecutionStrategyParameters` but the class is marked as `@Internal`);
- other small improvements: https://github.com/graphql-java/graphql-java/pull/3236/commits/dfb858dd4ee4acb04e4509875cfa531869dffaa3

D.

**Reply by bbakerman** [reply]
> t how about not using CompletableFuture in case of a TrivialDataFetcher? My hope is that since PropertyDataFetcher  is just doing in memory lookups, it does not really require async;

Sounds attractive right - but but we do more with CF than just fetch them.

We combine them all together and when do we "complete" the fields.  CFs also allow you to return 2 values.  The value itself AND hold whether an exception was thrown.  So when we fetch  / complete a field we can have exceptions happen.  If we passed back an `Object` (that could be either a CF or a direct value) then we cant hold any exceptions that might have happened as we did things

And with async code, a simple `try / catch` no longer works because the work to get / complete a value happens on different threads (in theory and sometimes in practice) so the try catch will not do its job easily.

The fact is when we made graphql-java async - we chose to use the CF pattern as our way to

* be async
* allow exceptions to be captured and propogated
* allow us to "compose together" a series of step

**Reply by bbakerman** [reply]
> NonNullableFieldValidator

I think we might be able to do something here - since its really a `executionContext +  executionStepInfo`.  We could maybe make this static - as long as there is not state being held in the form of some previous patent step implied here

> unfortunately, it is part of the public API... so it will be hard to change

I would be happy with such a breaking change - its not likely to be a key mechanism used by others outside our execution strategy code.  (If I\we had my time again we would not make `ExecutionStrategyParameters` API not the `ExecutionStrategy` to be honest

**Reply by bbakerman** [reply]
> ExecutionStrategyParameters has 2 fields that are not used anymore (listSize and currentListIndex): here it is hard to remove them because they are part of a public API class;

Again I am not sure these are crucial and unlikely to be used - removing them would be an acceptable breaking change

**Reply by bbakerman** [reply]
( we avoid breaking changes mostly but we also judge where things are really key code and hence likely to impact a large number of people)

**Reply by dfa1** [reply]
> > t how about not using CompletableFuture in case of a TrivialDataFetcher? My hope is that since PropertyDataFetcher  is just doing in memory lookups, it does not really require async;
> 
> Sounds attractive right - but but we do more with CF than just fetch them.
> 
> We combine them all together and when do we "complete" the fields. CFs also allow you to return 2 values. The value itself AND hold whether an exception was thrown. So when we fetch / complete a field we can have exceptions happen. If we passed back an `Object` (that could be either a CF or a direct value) then we cant hold any exceptions that might have happened as we did things
> 
> And with async code, a simple `try / catch` no longer works because the work to get / complete a value happens on different threads (in theory and sometimes in practice) so the try catch will not do its job easily.
> 
> The fact is when we made graphql-java async - we chose to use the CF pattern as our way to
> 
>     * be async
> 
>     * allow exceptions to be captured and propogated
> 
>     * allow us to "compose together" a series of step
thanks for the answer... it could be a FAQ or a blog post even :)


For the records, I tried the following "smart ideas":
- I actually tried to make the code "less async" for some special cases but it not simple and I still don't see any real gain, even for `PropertyDataFetcher`;
- merging two levels of nested futures in `AsyncExecutionStrategy#execute` but thanks to some written tests, the experiment failed really fast :)
- merging CF operations in `ExecutionStrategy#handle`
```
    return fetchedValue
                .handle((result, exception) -> {
                    fetchCtx.onCompleted(result, exception);
                    if (exception != null) {
                        return handleFetchingException(executionContext, dataFetchingEnvironment.get(), exception);
                    } else {
                        return CompletableFuture.completedFuture(result);
                    }
                })
                .thenCompose(Function.identity())
                .thenApply(result -> unboxPossibleDataFetcherResult(executionContext, parameters, result));
```
this is promising on paper but I could make it working properly yet;
- tried to optimize `ExecutionResultImpl` (as it is used a lot internally) to use `ImmutableList`... but I don't see any big gains yet;

**Reply by dfa1** [reply]
> > ExecutionStrategyParameters has 2 fields that are not used anymore (listSize and currentListIndex): here it is hard to remove them because they are part of a public API class;
> 
> Again I am not sure these are crucial and unlikely to be used - removing them would be an acceptable breaking change

I tried this quickly, the results are good:

baseline
Result "benchmark.BenchMark.benchMarkSimpleQueriesThroughput":
  62.101 ±(99.9%) 2.006 ops/s [Average]
  (min, avg, max) = (59.572, 62.101, 65.441), stdev = 1.877
  CI (99.9%): [60.095, 64.108] (assumes normal distribution)

after removal of currentIndex/listSize:

Result "benchmark.BenchMark.benchMarkSimpleQueriesThroughput":
  66.741 ±(99.9%) 4.272 ops/s [Average]
  (min, avg, max) = (62.991, 66.741, 75.928), stdev = 3.996
  CI (99.9%): [62.469, 71.014] (assumes normal distribution)
  
Basically, max throughput went from 65 to 75 and also average is a bit better. I'm going to double check the results before opening a PR ...

**Reply by bbakerman** [reply]
ps

https://github.com/graphql-java/graphql-java/pull/3235

This does one of the examples above

**Reply by dfa1** [reply]
> ps
> 
> 
> 
> https://github.com/graphql-java/graphql-java/pull/3235
> 
> 
> 
> This does one of the examples above

nice, one allocation down! But I also tried to fuse thenCompose and thenApply (still not working)

**Comment by rrva** [comment]
Great stuff! I have not yet benchmarked these changes on my workload but this work is welcome, thanks @dfa1! I implemented a custom execution strategy in kotlin using coroutines and no CompletableFutures at all, but currently it performs far worse than stock AsyncExecutionStrategy in throughput benchmarks. It does not support instrumentations at all. Can share that code but this main approach looks much more promising and has a better upgrade path etc etc.

**Comment by dfa1** [comment]
https://github.com/graphql-java/graphql-java/pull/3236 adding few more commits, please have a look when you have time
